### PR TITLE
LPS-115418 Add custom legend with checkboxes and stylish axis

### PR DESCRIPTION
**Pull request by dgarciasarai**
**Description**

The main goal of this pull request is to implement the ability to show/hide the bars in the Audit Graph. The data is mock for now until it is injected in the `*DisplayContext.java`.

**Changes**

* Add custom legend with checkboxes (the user can interact with the legend in the design)
* Change axis font size (our designer told me that should be 14)
* Transform Y-axis values to be shortened (to avoid large numbers that take a lot of space)
* Add style to legend checkboxes (bars are dynamic depending on the data, so the legend it's dynamic too).

**Feature flag: YES**

The Audit Graph is hidden. The configuration file to show the chart is here: https://issues.liferay.com/browse/LPS-115805

**Module**
```
~/liferay-portal/modules/apps/content-dashboard/content-dashboard-web
```

**Screenshots**

Screenshot 2020-06-24 at 12 11 15

Screenshot 2020-06-24 at 12 09 36

cc: @bryceosterhaus @wincent

### CODE REVIEW

**Comment by dgarciasarai** [comment]
ci:test:sf

**Comment by jbalsas** [comment]
Hey @dgarciasarai, thanks for sending this over and in particular for those lovely screenshots!!

On top of that... could you also provide some additional context as to:
- What does this PR do: it's not clear from the screenshots alone what's the change this brings.
- How do I test the changes: some of the reviewers might have the necessary context and some not. Even those that knew it might've forgotten about it, so putting it in writing would definitely help their task :)

We've encoded some of these things in our [Code Review Guidelines](https://github.com/liferay/liferay-frontend-guidelines/blob/master/dxp/code_review.md#frontend-infrastructure-team-code-review-process) and [Pull Request Etiquette](https://github.com/liferay/liferay-frontend-guidelines/blob/master/general/pull_requests.md). Please, take a look there (if you haven't already) to see the motivation behind these requests!

❤️

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/utils/numberShorten.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ * If a number is bigger than 10000 it will transform into thousands (K).
+ * If a number is bigger than 1000000 it will transform into millions (M).
+ * If a number is bigger than 1000000000 it will transform into billions (B).
+ *
+ * 4 => 4
+ * 4000 => '4K'
+ * 4000000 => '4M'
+ * 4000000000 => '4B'
+ */
+export function numberShorten(value) {
```
Comment (jbalsas) [inline modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/utils/numberShorten.js:25]: Did you try using [`formatStorage`](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/format_storage.es.js#L17)

Comment (jbalsas) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/utils/numberShorten.js:25]: One interesting distinction is that it has `denominator: 1024.0`... so either this or the other one might be slightly off 😉

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js [hunk 2]
```diff
@@ -81,13 +99,78 @@ export default function AuditBarChart({rtl, vocabularies}) {
...
+				}`;
+				});
+			}
+
+			head.appendChild(style);
+		}
+	}, [colors]);
+
+	useEffect(() => {
+		return () => {
```
Comment (jbalsas) [inline modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js:131]: Why don't you return this from the previous `useEffect`? This feels like that one's cleanup... 🤔

Comment (jbalsas) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js:131]: Also the fact that the other one has `[colors]` as dependencies and this one has none makes we wonder...

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/utils/numberShorten.js (hunk 1 above)
Comment (dgarciasarai) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/utils/numberShorten.js:25]: I didn't know that 🤔 I can configure the different suffix? 
`numberShorten` is used to represent the number with K (10.000 = 10K), M (1.000.000 = 1M), or B (1.000.000.000 = 1B). Can I get that with `formatStorage`? Thanks!

**Comment by dgarciasarai** [comment]
@jbalsas I added a long description with more context 😺❤️

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/utils/numberShorten.js (hunk 1 above)
Comment (jbalsas) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/utils/numberShorten.js:25]: [Tests will tell you that!](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-js/frontend-js-web/test/liferay/util/format_storage.es.js)

```javascript
	it('formats size under 1048575 bytes to kilobytes, with the default KB suffix', () => {
		expect(formatStorage(10400)).toEqual('10KB');
	});

	it('formats size 0 bytes to kilobytes, with the default KB suffix', () => {
		expect(formatStorage(0)).toEqual('0KB');
	});

	it('formats size over 1048575 bytes to megabytes, with the default MB suffix', () => {
		expect(formatStorage(1048576)).toEqual('1.0MB');
	});

	it('formats size over 1048575 bytes to megabytes with custom space, decimal separator, and suffix type parameters', () => {
		expect(
			formatStorage(1048576, {
				addSpaceBeforeSuffix: true,
				decimalSeparator: ',',
				suffixMB: 'megabytes',
			})
		).toEqual('1,0 megabytes');
	});
```

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js (hunk 2 above)
Comment (dgarciasarai) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js:129]: Done 👍

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/utils/numberShorten.js (hunk 1 above)
Comment (dgarciasarai) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/utils/numberShorten.js:25]: I don't know if `formatStorage` is the perfect method for my use case... (see Y-Axis) 🤔 

Screenshot 2020-06-24 at 18 16 41

I used:

```
Liferay.Util.formatStorage(payload.value, {
	denominator: 1000,
	suffixGB: 'B',
	suffixKB: 'K',
	suffixMB: 'M',
})
```
What do you think @jbalsas?

Comment (jbalsas) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/utils/numberShorten.js:25]: So, these are views, so you're grouping per-1000 views? or something like that?

What's your mocked data for the `YAxis`? Based on your original screenshot, it should maybe be `denominator: 10`?

Comment (dgarciasarai) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/utils/numberShorten.js:25]: The Y-Axis values are automatically calculated by Recharts, I am not adding it manually. I mocked the values of the bars:
```
const vocabularies = [
	{
		categories: [
			{
				key: 'education',
				name: 'Education',
				value: 478,
				vocabularyName: 'Stage',
			},
			{
				key: 'selection',
				name: 'Selection',
				value: 1055,
				vocabularyName: 'Stage',
			},
			{
				key: 'selection',
				name: 'Selection',
				value: 822,
				vocabularyName: 'Stage',
			},
		],
		key: 'business-decision-maker',
		name: 'Business Decision Maker',
		vocabularyName: 'Audience',
	},
	{ ... }
];
```

Comment (jbalsas) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/utils/numberShorten.js:25]: That's what I mean, feels like you have `[500, 1000, 1500, 2000, 2500]` which `/1000` => `[0, 1, 1, 2, 2]`, so all is good there.

Comment (jbalsas) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/utils/numberShorten.js:25]: But yeah, no need to worry about this if it's not clear...

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js [hunk 3]
```diff
@@ -68,10 +70,26 @@ export default function AuditBarChart({rtl, vocabularies}) {
...
+			{}
+		);
+
+		const colors = bars.reduce(
+			(acc, {dataKey}, index) => ({
+				...acc,
+				[dataKey]: COLORS[index % COLORS.length],
+			}),
+			{}
+		);
```
Comment (bryceosterhaus) [inline modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js:87]: We could combine these into a single reduce instead of doing it twice, you can do something like


```js
const {legendCheckboxes, colors} = bars.reduce(
	(acc, {dataKey}, index) => ({
		colors: {
			...acc.colors,
			[dataKey]: COLORS[index % COLORS.length],
		},
		legendCheckboxes: {
			...acc. legendCheckboxes,
			[dataKey]: true,
		},
	}),
	{legendCheckboxes: {}, colors: {}}
);
```

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js [hunk 4]
```diff
@@ -81,13 +99,76 @@ export default function AuditBarChart({rtl, vocabularies}) {
...
+
+	useEffect(() => {
+		if (Object.keys(colors).length) {
+			const head =
+				document.head || document.getElementsByTagName('head')[0];
+			const style = document.createElement('style');
+			style.id = 'legend-checkboxes';
+			style.type = 'text/css';
+
+			{
```
Comment (bryceosterhaus) [inline modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js:112]: What are these braces for?

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js [hunk 5]
```diff
@@ -81,13 +99,76 @@ export default function AuditBarChart({rtl, vocabularies}) {
...
 		: [BAR_CHART.legendHeight + BAR_CHART.dotRadiusMin];
 
+	const [checkboxes, setCheckbox] = useState(legendCheckboxes);
+
+	useEffect(() => {
+		if (Object.keys(colors).length) {
+			const head =
+				document.head || document.getElementsByTagName('head')[0];
+			const style = document.createElement('style');
+			style.id = 'legend-checkboxes';
```
Comment (bryceosterhaus) [inline modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js:109]: Is it possible that there will be multiple `AuditBarChart` components on the page? If so, we run the risk of having duplicate ids on the same page by hard coding this id here.

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js [hunk 6]
```diff
@@ -81,13 +99,76 @@ export default function AuditBarChart({rtl, vocabularies}) {
...
+				head.removeChild(style);
+			}
+		};
+	}, [colors]);
+
+	const renderLegend = (props) => {
+		const {payload} = props;
+
+		return (
+			<>
```
Comment (bryceosterhaus) [inline modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js:143]: Is this fragment necessary? I think you can return the `.map` without having to wrap it.

```jsx
() => {
    return payload.map(entry => ...)
}
```

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js (hunk 3 above)
Comment (dgarciasarai) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js:87]: That's right!

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js (hunk 4 above)
Comment (dgarciasarai) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js:112]: Not needed 👍

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js (hunk 6 above)
Comment (dgarciasarai) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js:143]: That's right! Thanks!

**Comment by dgarciasarai** [comment]
ci:test:sf

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js (hunk 5 above)
Comment (jbalsas) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js:107]: Are you sure? What about tomorrow? 

This feels a little like our other conversation last week regarding the mocked data... consider the cost of some fixes and their future risk. Sometimes a fix is simple enough and can save you enough headaches in the future (when you or someone else wants to use this component in a different way than you're expecting) that it's worth the extra time.

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js [hunk 7]
```diff
@@ -81,13 +97,70 @@ export default function AuditBarChart({rtl, vocabularies}) {
 		? [BAR_CHART.dotRadiusMin]
 		: [BAR_CHART.legendHeight + BAR_CHART.dotRadiusMin];
 
+	const [checkboxes, setCheckbox] = useState(legendCheckboxes);
+
+	useEffect(() => {
+		if (Object.keys(colors).length) {
+			const head =
+				document.head || document.getElementsByTagName('head')[0];
```
Comment (wincent) [inline modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js:105]: Why not just `document.head`?

[TypeScript thinks it is guaranteed to be non-null](https://github.com/microsoft/TypeScript/blob/cc7cb9eadb10a56687513d4fc3a67d5210b387bb/src/lib/dom.generated.d.ts#L4505), and [MDN doesn't raise any red flags](https://developer.mozilla.org/en-US/docs/Web/API/Document/head) in the browser compatibility section.

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js [hunk 8]
```diff
@@ -81,13 +97,70 @@ export default function AuditBarChart({rtl, vocabularies}) {
...
+			Object.entries(colors).forEach(([dataKey, color]) => {
+				style.textContent += `.custom-control-color-${dataKey}.custom-control-input:checked ~ 
+					.custom-control-label::before {
+						background-color: ${color};
+						border-color: ${color};
+				}
+				.custom-control-color-${dataKey}.custom-control-input:not(:checked) ~ 
+					.custom-control-label::before {
+						border-color: ${color};
+				}`;
```
Comment (wincent) [inline modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js:119]: style nit: this might look nicer:

```
style.textContent += `
	.custom-control-color-${dataKey}.custom-control-input:checked ~ 
		.custom-control-label::before {
			background-color: ${color};
			border-color: ${color};
	}
	.custom-control-color-${dataKey}.custom-control-input:not(:checked) ~ 
		.custom-control-label::before {
			border-color: ${color};
	}
`;
```

(Not just nicer in the source here, but when you inspect the DOM, because you'll have a newline between each pair instead of them all running together.)

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js [hunk 9]
```diff
@@ -81,13 +97,70 @@ export default function AuditBarChart({rtl, vocabularies}) {
...
+			});
```
Comment (wincent) [inline modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js:120]: To avoid repeatedly accessing (reading then overwriting) the `.textContent` property, you could `reduce()` here too. Warning: this is almost certainly pointless microoptimization, but I like the way it puts the interesting information right up front (ie. before I even get to the `reduce()` I can see that the purpose of all this and the end result is going to be an assignment to `style.textContent`).

```js
if (Object.keys(colors).length) {
	const head = document.head || document.getElementsByTagName("head")[0];
	const style = document.createElement("style");
	style.id = "legend-checkboxes";

	style.type = "text/css";

	style.textContent = Object.entries(colors).reduce(
		(acc, [dataKey, color]) => {
			return acc.concat(`
				.custom-control-color-${dataKey}.custom-control-input:checked ~ 
					.custom-control-label::before {
						background-color: ${color};
						border-color: ${color};
				}
				.custom-control-color-${dataKey}.custom-control-input:not(:checked) ~ 
					.custom-control-label::before {
						border-color: ${color};
				}
			`);
		}
	);

	head.appendChild(style);
}
```

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/utils/numberShorten.js (hunk 1 above)
Comment (wincent) [inline modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/utils/numberShorten.js:25]: nit: I would expect this to be `shortenNumber` (ie. verb-object, not object-verb).

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/utils/numberShorten.js [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+	if (value >= 1000000000) {
```
Comment (wincent) [inline modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/utils/numberShorten.js:26]: For readability and to reduce the chances of making a mistake, you might be able to write `1_000_000_000` here (check though; off the top of my head, I don't think we have [babel-plugin-proposal-numeric-separator](https://babeljs.io/docs/en/babel-plugin-proposal-numeric-separator) explicitly enabled, even though we could because [the proposal](https://github.com/tc39/proposal-numeric-separator) is at stage 3; but it's something we should definitely enable).

I created https://github.com/liferay/liferay-npm-tools/issues/470 as a reminder to turn this plugin on by default, if it isn't already.

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/utils/numberShorten.js [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+		return value / 1000000000 + 'B';
```
Comment (wincent) [inline modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/utils/numberShorten.js:27]: So all these are going to return floats. It's not really going to "shorten" the number at all.

eg. `4298123471 / 1000000000` is `4.298123471`, and if you add `B` to that, you've actually made it longer, not shorter.

Also, you're concatenating a number and a string, and TypeScript will yell at you for that, so it would be better to avoid the implicit coercion.

Did you consider something like?

```js
return value / 1000000000).toFixed(0) + 'B';
```

which is a nicer way to write:

```js

Math.floor(4298123471 / 1000000000).toString() + 'B';
```

tl;dr: I think you should make this function actually shorten the number, or if that is not the intended behavior, rename the function to describe what it does.

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js (hunk 7 above)
Comment (dgarciasarai) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js:105]: Ok, I'll change it. Thanks!

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js [hunk 12]
```diff
@@ -81,13 +97,70 @@ export default function AuditBarChart({rtl, vocabularies}) {
 		? [BAR_CHART.dotRadiusMin]
 		: [BAR_CHART.legendHeight + BAR_CHART.dotRadiusMin];
 
+	const [checkboxes, setCheckbox] = useState(legendCheckboxes);
+
+	useEffect(() => {
```
Comment (wincent) [inline modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js:102]: One thing to consider is that [`useLayoutEffect`](https://reactjs.org/docs/hooks-reference.html#uselayouteffect) *might* be a better choice here. It probably doesn't matter because I don't think this component is likely to update often, but if it did, there is a chance that `useLayoutEffect` might avoid a repaint, because:

> Updates scheduled inside `useLayoutEffect` will be flushed synchronously, before the browser has a chance to paint.

In other words, you update the chart DOM, you inject the `style` element, and *then* the browser repaints (once), instead of updating chart, layout/painting, injecting style element, repainting. But like I said, unless the component is rendering in a loop, you are unlikely to notice a difference. Just wanted to bring it up though as food for thought.

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js (hunk 8 above)
Comment (dgarciasarai) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js:119]: 👍

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js (hunk 9 above)
Comment (dgarciasarai) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js:120]: I ❤️ easy reading

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/utils/numberShorten.js (hunk 1 above)
Comment (dgarciasarai) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/utils/numberShorten.js:25]: 👍

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/utils/numberShorten.js (hunk 10 above)
Comment (dgarciasarai) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/utils/numberShorten.js:26]: I tried that but I have a babel error. Should I enable something to make this change or should I wait until it' definitely enabled? 🤔

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/utils/numberShorten.js (hunk 11 above)
Comment (dgarciasarai) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/utils/numberShorten.js:27]: I'm not controlling the `value` that the function is receiving. It's automatically calculated by Recharts to display the ticks in the chart. I think I'm going with this to prevent longer numbers instead of shortening:

```
return value / 1000000000).toFixed(0) + 'B';
```

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/utils/numberShorten.js (hunk 10 above)
Comment (wincent) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/utils/numberShorten.js:26]: Yeah, just leave it and one day it will magically start working. When we add it we will do a grep (`git grep '\b\d{5,}' -- '*.js'`) looking for long-ish numbers to update to use the new syntax.

Comment (dgarciasarai) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/utils/numberShorten.js:26]: Great! Thanks! 😄

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/utils/numberShorten.js (hunk 11 above)
Comment (wincent) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/utils/numberShorten.js:27]: Sounds good.

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js (hunk 12 above)
Comment (dgarciasarai) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js:102]: I'm controlling the show/hide bars with the checkbox in the legend with a state in the component but probably in the future, we want that when the user filters the list, affects also in the chart and for that reason maybe `useLayoutEffect` it's a better choice. What do you think?

Comment (wincent) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js:102]: By "frequent updates" I mean something like I said above (eg. "rendering in a loop", something that could happen many times a second, while dragging or scrolling or typing, for instance); something like toggling a checkbox is going to be very infrequent, and limited by the user's ability to click (how many times can you toggle a checkbox in a second? Five?).

I would say if you don't see any visible flicker on render, don't worry about `useLayoutEffect` yet.

Comment (dgarciasarai) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js:102]: 👍

Comment (jbalsas) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js:102]: > (how many times can you toggle a checkbox in a second? Five?)

I dare you... I double dare you...

[My Score: 12 CPS](https://clickspeedtest.com/clicks-per-second.html)

just-sayin

**Comment by dgarciasarai** [comment]
@wincent I updated the pull request with your feedback. Thanks!

**Comment by dgarciasarai** [comment]
ci:test:sf

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/utils/shortenNumber.js [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+		return (value / 1000000000).toFixed(0) + 'B';
+	}
+	else if (value >= 1000000) {
+		return (value / 1000000).toFixed(0) + 'M';
+	}
+	else if (value >= 10000) {
+		return (value / 10000).toFixed(0) + 'K';
+	}
+
+	return value;
```
Comment (wincent) [inline modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/utils/shortenNumber.js:36]: I know you're not going for a fully generic implementation for a "shorten" function but if you were ever to call this function with a number like `4.98724243424`, you wouldn't get something very short back. Up to you whether you want to do anything about it. Could be as simple as `return Math.floor(value)` or something else to your taste.

**Comment by dgarciasarai** [comment]
Done! @bryceosterhaus @jbalsas @wincent 😉

**Comment by dgarciasarai** [comment]
ci:test:sf

**Comment by bryceosterhaus** [comment]
looks good!

**Comment by bryceosterhaus** [comment]
ci:forward
